src/rust/lib_ccxr/src/util/net/common.rs

```diff
@@ -0,0 +1,353 @@
+    pub fn bin_data(data: &'a [u8]) -> Block<'a> {
+        Block::new(Command::BinData, data)
+    }
+
+    /// Create a new [`Password`](Command::Password) Block along with `password` data.
+    ///
+    /// The data of the returned [`Block`] will consist of `password` encoded as UTF-8 bytes which  
+    /// is not nul-terminated.
+    ///
+    /// # Examples
+    /// ```
+    /// # use lib_ccxr::util::net::Block;
+    /// let b = Block::password("A");
+    /// assert_eq!(b.data(), &[b'A']);
+    /// ```
+    pub fn password(password: &'a str) -> Block<'a> {
+        Block::new(Command::Password, password.as_bytes())
+    }
+
+    /// Create a new [`CcDesc`](Command::CcDesc) Block along with `desc` data.
+    ///
+    /// The data of the returned [`Block`] will consist of `desc` encoded as UTF-8 bytes which is
+    /// not nul-terminated.
+    ///
+    /// # Examples
+    /// ```
+    /// # use lib_ccxr::util::net::Block;
+    /// let b = Block::cc_desc("Teletext");
+    /// assert_eq!(b.data(), &[b'T', b'e', b'l', b'e', b't', b'e', b'x', b't']);
+    /// ```
+    pub fn cc_desc(desc: &'a str) -> Block<'a> {
+        Block::new(Command::CcDesc, desc.as_bytes())
+    }
+
+    /// Create a new [`EpgData`](Command::EpgData) Block along with the related metadata used in
+    /// EPG.
+    ///
+    /// All the parameters are encoded as UTF-8 bytes which are nul-terminated. If a parameter is
+    /// [`None`], then it is considered to be equivalent to an empty String. All these
+    /// nul-terminated UTF-8 bytes are placed one after the other in the order of `start`, `stop`,
+    /// `title`, `desc`, `lang`, `category` with nul character acting as the seperator between
+    /// these sections.
+    ///
+    /// # Examples
+    /// ```
+    /// # use lib_ccxr::util::net::Block;
+    /// let b = Block::epg_data("A", "B", Some("C"), None, Some("D"), None);
+    /// assert_eq!(b.data(), &[b'A', b'\0', b'B', b'\0', b'C', b'\0', b'\0', b'D', b'\0', b'\0']);
+    /// ```
+    pub fn epg_data(
+        start: &str,
+        stop: &str,
+        title: Option<&str>,
+        desc: Option<&str>,
+        lang: Option<&str>,
+        category: Option<&str>,
+    ) -> Block<'a> {
+        let title = title.unwrap_or("");
+        let desc = desc.unwrap_or("");
+        let lang = lang.unwrap_or("");
+        let category = category.unwrap_or("");
+
+        // Plus 1 to accomodate space for the nul character
+        let start_len = start.len() + 1;
+        let stop_len = stop.len() + 1;
+        let title_len = title.len() + 1;
+        let desc_len = desc.len() + 1;
+        let lang_len = lang.len() + 1;
+        let category_len = category.len() + 1;
+
+        let total_len = start_len + stop_len + title_len + desc_len + lang_len + category_len;
+        let mut data = Vec::with_capacity(total_len);
+
+        data.extend_from_slice(start.as_bytes());
+        data.extend_from_slice("\0".as_bytes());
```

Comment:
Is this necessary?? @elbertronnie

Reply:
Is this because C code uses `memcpy`. But still is this really needed?

Reply:
```c
t = 1;
if (title != NULL)
    t += strlen(title);

// ------- other code -------

if (title != NULL)
    memcpy(end, title, t);
end += t;
```
Yes, because `memcpy` will also also copy the `NUL` character. This is necessary, otherwise how will the program know where one string starts and other ends. You are not passing the length of each string anyway.

Reply:
And I was busy for the past week, I will probably review both your PRs tomorrow or day after tomorrow.